src/messaging/ExchangeContext.cpp

```diff
@@ -108,10 +135,34 @@ CHIP_ERROR ExchangeContext::SendMessage(uint16_t protocolId, uint8_t msgType, Pa
         }
     }
 
-    payloadHeader.SetInitiator(IsInitiator());
+    // Send the message.
+    if (payloadHeader.IsNeedsAck())
+    {
+        ReliableMessageManager::RetransTableEntry * entry = nullptr;
 
-    err = mExchangeMgr->GetSessionMgr()->SendMessage(payloadHeader, mPeerNodeId, std::move(msgBuf));
-    SuccessOrExit(err);
+        // Add to Table for subsequent sending
+        err =
+            mExchangeMgr->GetReliableMessageMgr()->AddToRetransTable(&mReliableMessageContext, payloadHeader, mPeerNodeId, &entry);
+        SuccessOrExit(err);
+
+        err = mExchangeMgr->GetSessionMgr()->SendMessage(payloadHeader, mPeerNodeId, std::move(msgBuf), &entry->retainedBuf);
+
+        if (err != CHIP_NO_ERROR)
+        {
+            // Remove from table
+            ChipLogError(ExchangeManager, "Failed to send message to 0x%lx with err %ld", mPeerNodeId, long(err));
+            mExchangeMgr->GetReliableMessageMgr()->ClearRetransTable(*entry);
+        }
+        else
+        {
+            mExchangeMgr->GetReliableMessageMgr()->StartFromRetransTable(entry);
+        }
+    }
+    else
+    {
+        err = mExchangeMgr->GetSessionMgr()->SendMessage(payloadHeader, mPeerNodeId, std::move(msgBuf));
+        SuccessOrExit(err);
+    }
```

Comment:
Same as above.

Would you please extract these logics into a function in `ReliableMessageContext` or `ReliableMessageManager` class ?

The code is totally fine, and reasonable. But putting the code here will significantly distract one's mind when reading the code.

Reply:
This logic is the original logic for non-CRMP case if no acking is required from sender.